kong/plugins/ddtrace/sampler.lua

```diff
@@ -137,30 +137,20 @@ local function apply_initial_sample_rate(sampler, span)
     return false
 end
 
-
-local function apply_agent_sample_rate(sampler, span) 
-    local service = span.service
-    if not service then
-        service = ""
-    end
-    local env = span.meta["env"]
-    if not env then
-        env = ""
-    end
+local function apply_agent_sample_rate(sampler, span)
+    local service = span.service or ""
+    local env = span.meta["env"] or ""
 
     local sample_rate_key = "service:" .. service .. ",env:" .. env
-    local service_env = sampler.agent_sample_rates[sample_rate_key]
-    if service_env then
-        local sampled = sampling_decision(span, service_env.max_id)
-        span.metrics["_dd.agent_psr"] = service_env.rate
-        return true, sampled
-    end
-    local default = sampler.agent_sample_rates[default_sampling_rate_key]
-    if default then
-        local sampled = sampling_decision(span, default.max_id)
-        span.metrics["_dd.agent_psr"] = default.rate
+
+    local rule = sampler.agent_sample_rates[sample_rate_key]
+    if rule then
+        local sampled = sampling_decision(span, rule.max_id)
+        span.meta["_dd.p.dm"] = "-1" -- "AGENT RATE"
+        span.metrics["_dd.agent_psr"] = rule.rate
         return true, sampled
```

Comment:
@dgoffredo I removed the default here, not sure about this one. I will revisit the sampler RFC on Monday otherwise.

Reply:
There are two notions of "default." The first is in `default_sampling_rate_key`, which means "I've received a 'catch-all' pattern from the Datadog Agent, so if no other "service:...,env:..." entry matches, I can use the rate associated with the default, i.e. key 'service:,env:'". [doc][1]. This "default," together with any non-default Agent-provided rates, are associated with sampling mechanism 1.

The other notion of "default" only happens before we've received the "service:,env:" entry in a response from the Agent, i.e. within the first ~10 seconds of tracing. In that case, we might have literally nothing to go on, and so the decision is to keep everything, and the associated sampling mechanism is 0.

[1]: https://github.com/DataDog/datadog-agent/blob/d2899451c5e7e8d205cdaf699fa9879980891e09/pkg/trace/api/version.go#L30

Reply:
Thank you for those details. If I understood correctly, I have to remove the [default_sampling_rate_key.](https://github.com/DataDog/kong-plugin-ddtrace/blob/b5a52bd95a236a0ddf243fe15568e6be256820aa/kong/plugins/ddtrace/sampler.lua#L80)

Reply:
Here is my interpretation: 16a59999ccda0fca79f0ff63b36b22f1a7044c12. We can talk about it during our daily meeting.